Fix text color not resolving when `CupertinoThemeData.brightness` is null

ivirtex opened the pull request:
In the build method of the `CupertinoApp`, `CupertinoTheme.of(context)` gets called before MediaQuery exists in the tree.

By default, every value of the `CupertinoTheme.data` gets resolved according to the `CupertinoThemeData.brightness`, but if `CupertinoThemeData.brightness` is null, `MediaQuery.maybeOf(context)?.platformBrightness` is used.
Due to the lack of `MediaQuery` in the tree, `CupertinoTheme.data` gets its values resolved against `Brightness.light`.
This causes text color to be constantly equal to black - `0xff000000`.

https://github.com/flutter/flutter/blob/8b32ac7a512b35b2b01da949ff41be814d025207/packages/flutter/lib/src/cupertino/colors.dart#L990-L994
Fixes #48438

Before / After
<img src="https://user-images.githubusercontent.com/57679865/200957641-ebd8ad75-a06e-49be-8e87-10f5baf281a1.png" width="40%"/>  <img src="https://user-images.githubusercontent.com/57679865/200957649-6175c8e1-0e10-4eff-b90c-a12d1b91a2e6.png" width="40%"/>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ivirtex committed:
Fix text color not resolving in `CupertinoApp`

ivirtex committed:
Make linter happy

ivirtex commented:
> Golden file changes have been found for this pull request. Click [here to view and triage](https://flutter-gold.skia.org/cl/github/115026) (e.g. because this is an intentional change).

Seems like this is related to the https://github.com/flutter/flutter/pull/114450.
After merging the revert, goldens are passing.

ivirtex commented:
Hey @MitchellGoodwin, I've seen that you requested a review of this PR yourself almost 1.5 months ago, and I just wanted to remind you about that in case you lost my PR in your review queue.

justinmc reviewed (commented):
I posted a workaround and some details on the issue: https://github.com/flutter/flutter/issues/48438#issuecomment-1361908926

Ideally a solution would make Material and Cupertino behave identically here, where the `brightness` is set automatically on the theme data, and they would both do so in the same way.

MitchellGoodwin commented:
Sorry I haven't responded to this sooner. I've been keeping an eye on this PR, but because it's dealing with code at the app level, I wanted to get a second opinion before reviewing.

I think the reason why Material works correctly but Cupertino doesn't is that the check for the platform brightness happens in a builder passed to MaterialApp.

https://github.com/flutter/flutter/blob/9a347fb06366632d236459a0ea972ffda36e2acc/packages/flutter/lib/src/material/app.dart#L917

But in Cupertino it's happening in a static method in CupertinoTheme, so it's not automatically updating when the MediaQuery enters the tree. I'm guessing your solution is indirectly solving the issue by causing a rebuild in CupertinoApp that catches the change in the static method. Ideally the way the theme gets the default brightness would be passed in similar to Material in the builder.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/app.dart:
```diff
@@ -609,9 +609,15 @@ class _CupertinoAppState extends State<CupertinoApp> {
-                builder: _buildWidgetApp,
-              ),
+              child: !widget.useInheritedMediaQuery
+                  ? MediaQuery.fromWindow(
```
It looks like `MediaQuery.fromWindow` is planed for deprecation as well. https://github.com/flutter/flutter/pull/117480

ivirtex committed:
Create cupertinoBuilder for theming

ivirtex committed:
Make linter happy

ivirtex committed:
Fix tests

ivirtex commented:
I've removed `MediaQuery` workaround and created `_cupertinoBuilder` similar to the `_materialBuilder` in the `MaterialApp`. In order for text to resolve, `DefaultTextStyle` must be declared in the builder rather than passing a `textStyle` to the `WidgetsApp`.

It works great, but I also found something else, `color` passed to the `WidgetsApp` is not being resolved according to the platform brightness, but only to the `brightness` property of the `CupertinoThemeData` (the same situation as with the text).
I think there are 2 options here:
1. Leave it as is, let it resolve only to the manually set brightness and end up with 2 `CupertinoTheme` widgets in the tree (one from the `_cupertinoBuilder` and one from the `build` method of `CupertinoApp`).
2. Pass color unresolved, similarly to the `MaterialApp` (no real use in iOS either?):
https://github.com/flutter/flutter/blob/7ddf42eae5ee9b9d770ce40aae362b0fc2ca1005/packages/flutter/lib/src/material/app.dart#L966-L974

MitchellGoodwin commented:
Thanks for making this update. Is there no way in either builder to get the platform brightness in order to get the right colors? Either through MediaQuery, or the workaround Justin mentioned?

ivirtex committed:
Fix regression

ivirtex committed:
Delete whitespace

ivirtex committed:
Resolve color passed to the `WidgetsApp`

ivirtex committed:
Revert "Resolve color passed to the `WidgetsApp`"

This reverts commit 6994719a564c476dcb6d389e4547bff3ded08203.

ivirtex commented:
Well, it would be technically possible using the `WidgetsBindingObserver` Justin mentioned, but it would look like reimplementing `CupertinoDynamicColor`(only for resolving one color!), colors need to be resolved not only by platform brightness, but also by accessibility features e.g. high contrast.
Is `color` property of `WidgetsApp` even used on iOS? From what I tested, there is no visual difference when this color is resolved or not.

ivirtex commented:
It can also be done by inserting a `MediaQuery` above the `_buildWidgetApp` builder, but we would have to wait for the https://github.com/flutter/flutter/pull/118004 to land.

MitchellGoodwin commented:
> It can also be done by inserting a `MediaQuery` above the `_buildWidgetApp` builder, but we would have to wait for the #118004 to land.

I think this makes the most sense to me, and we can fix the original issue, and move this one to a new one, if that is taking a while.

ivirtex committed:
Resolve selection colors

ivirtex committed:
Kick tests

ivirtex commented:
Alright, it seems that https://github.com/flutter/flutter/pull/116924 and https://github.com/flutter/flutter/pull/118004 accidentally fixed the original bug with text color and it is now correctly resolved (`View`, and therefore `MediaQuery`, is higher in the tree than `CupertinoTheme` so it can depend on it and grab correct platform brightness).

I've also noticed that `DefaultSelectionStyle` has its colors unresolved, so I fixed that and added relevant tests.

ivirtex committed:
Delete doubled test

MitchellGoodwin commented:
My apologies @ivirtex, this somewhat slipped my attention. Can you rebase on the latest? It looks like I can't with the new option due to conflicts. Otherwise this looks gtg to me.

ivirtex commented:
@MitchellGoodwin Conflicts are now resolved.

Btw, would you mind taking a look at https://github.com/flutter/platform_tests/pull/18?
